{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[NbConvertApp] WARNING | pattern 'config_template.ipynb' matched no files\n",
      "This application is used to convert notebook files (*.ipynb) to various other\n",
      "formats.\n",
      "\n",
      "WARNING: THE COMMANDLINE INTERFACE MAY CHANGE IN FUTURE RELEASES.\n",
      "\n",
      "Options\n",
      "-------\n",
      "\n",
      "Arguments that take values are actually convenience aliases to full\n",
      "Configurables, whose aliases are listed on the help line. For more information\n",
      "on full configurables, see '--help-all'.\n",
      "\n",
      "--debug\n",
      "    set log level to logging.DEBUG (maximize logging output)\n",
      "--generate-config\n",
      "    generate default config file\n",
      "-y\n",
      "    Answer yes to any questions instead of prompting.\n",
      "--execute\n",
      "    Execute the notebook prior to export.\n",
      "--allow-errors\n",
      "    Continue notebook execution even if one of the cells throws an error and include the error message in the cell output (the default behaviour is to abort conversion). This flag is only relevant if '--execute' was specified, too.\n",
      "--stdin\n",
      "    read a single notebook file from stdin. Write the resulting notebook with default basename 'notebook.*'\n",
      "--stdout\n",
      "    Write notebook output to stdout instead of files.\n",
      "--inplace\n",
      "    Run nbconvert in place, overwriting the existing notebook (only \n",
      "    relevant when converting to notebook format)\n",
      "--clear-output\n",
      "    Clear output of current file and save in place, \n",
      "    overwriting the existing notebook.\n",
      "--no-prompt\n",
      "    Exclude input and output prompts from converted document.\n",
      "--log-level=<Enum> (Application.log_level)\n",
      "    Default: 30\n",
      "    Choices: (0, 10, 20, 30, 40, 50, 'DEBUG', 'INFO', 'WARN', 'ERROR', 'CRITICAL')\n",
      "    Set the log level by value or name.\n",
      "--config=<Unicode> (JupyterApp.config_file)\n",
      "    Default: ''\n",
      "    Full path of a config file.\n",
      "--to=<Unicode> (NbConvertApp.export_format)\n",
      "    Default: 'html'\n",
      "    The export format to be used, either one of the built-in formats, or a\n",
      "    dotted object name that represents the import path for an `Exporter` class\n",
      "--template=<Unicode> (TemplateExporter.template_file)\n",
      "    Default: ''\n",
      "    Name of the template file to use\n",
      "--writer=<DottedObjectName> (NbConvertApp.writer_class)\n",
      "    Default: 'FilesWriter'\n",
      "    Writer class used to write the  results of the conversion\n",
      "--post=<DottedOrNone> (NbConvertApp.postprocessor_class)\n",
      "    Default: ''\n",
      "    PostProcessor class used to write the results of the conversion\n",
      "--output=<Unicode> (NbConvertApp.output_base)\n",
      "    Default: ''\n",
      "    overwrite base name use for output files. can only be used when converting\n",
      "    one notebook at a time.\n",
      "--output-dir=<Unicode> (FilesWriter.build_directory)\n",
      "    Default: ''\n",
      "    Directory to write output(s) to. Defaults to output to the directory of each\n",
      "    notebook. To recover previous default behaviour (outputting to the current\n",
      "    working directory) use . as the flag value.\n",
      "--reveal-prefix=<Unicode> (SlidesExporter.reveal_url_prefix)\n",
      "    Default: ''\n",
      "    The URL prefix for reveal.js. This can be a a relative URL for a local copy\n",
      "    of reveal.js, or point to a CDN.\n",
      "    For speaker notes to work, a local reveal.js prefix must be used.\n",
      "--nbformat=<Enum> (NotebookExporter.nbformat_version)\n",
      "    Default: 4\n",
      "    Choices: [1, 2, 3, 4]\n",
      "    The nbformat version to write. Use this to downgrade notebooks.\n",
      "\n",
      "To see all available configurables, use `--help-all`\n",
      "\n",
      "Examples\n",
      "--------\n",
      "\n",
      "    The simplest way to use nbconvert is\n",
      "    \n",
      "    > jupyter nbconvert mynotebook.ipynb\n",
      "    \n",
      "    which will convert mynotebook.ipynb to the default format (probably HTML).\n",
      "    \n",
      "    You can specify the export format with `--to`.\n",
      "    Options include ['asciidoc', 'custom', 'html', 'latex', 'markdown', 'notebook', 'pdf', 'python', 'rst', 'script', 'slides']\n",
      "    \n",
      "    > jupyter nbconvert --to latex mynotebook.ipynb\n",
      "    \n",
      "    Both HTML and LaTeX support multiple output templates. LaTeX includes\n",
      "    'base', 'article' and 'report'.  HTML includes 'basic' and 'full'. You\n",
      "    can specify the flavor of the format used.\n",
      "    \n",
      "    > jupyter nbconvert --to html --template basic mynotebook.ipynb\n",
      "    \n",
      "    You can also pipe the output to stdout, rather than a file\n",
      "    \n",
      "    > jupyter nbconvert mynotebook.ipynb --stdout\n",
      "    \n",
      "    PDF is generated via latex\n",
      "    \n",
      "    > jupyter nbconvert mynotebook.ipynb --to pdf\n",
      "    \n",
      "    You can get (and serve) a Reveal.js-powered slideshow\n",
      "    \n",
      "    > jupyter nbconvert myslides.ipynb --to slides --post serve\n",
      "    \n",
      "    Multiple notebooks can be given at the command line in a couple of \n",
      "    different ways:\n",
      "    \n",
      "    > jupyter nbconvert notebook*.ipynb\n",
      "    > jupyter nbconvert notebook1.ipynb notebook2.ipynb\n",
      "    \n",
      "    or you can specify the notebooks list in a config file, containing::\n",
      "    \n",
      "        c.NbConvertApp.notebooks = [\"my_notebook.ipynb\"]\n",
      "    \n",
      "    > jupyter nbconvert --config mycfg.py\n",
      "\n",
      "{'GAU': 'A', 'GGC': 'D', 'GCA': 'L', 'GUG': 'Y', 'UAU': 'K', 'UGC': 'Q', 'UCA': 'H', 'UUG': 'F', 'AAU': 'R', 'AGC': 'I', 'ACA': 'T', 'AUG': 'M', 'CAU': 'W', 'CGC': 'G', 'CCA': 'V', 'CUG': 'S', 'AAG': '*', 'GGG': '*', 'CUC': '*', 'CAG': '*', 'GUU': '*', 'UUC': '*', 'UCU': '*', 'CAA': '*', 'UUU': '*', 'UGU': '*', 'UAG': '*', 'ACU': '*', 'AGG': '*', 'AGA': '*', 'CCU': '*', 'AGU': '*', 'GAC': '*', 'GGA': '*', 'ACG': '*', 'CGU': '*', 'GUA': '*', 'CAC': '*', 'UGG': '*', 'AUA': '*', 'CCC': '*', 'ACC': '*', 'UGA': '*', 'AAA': '*', 'CUA': '*', 'GUC': '*', 'GAG': '*', 'UUA': '*', 'UCG': '*', 'GGU': '*', 'CGA': '*', 'AUC': '*', 'UCC': '*', 'CUU': '*', 'GCU': '*', 'UAA': '*', 'AAC': '*', 'GAA': '*', 'CGG': '*', 'CCG': '*', 'GCG': '*', 'GCC': '*', 'UAC': '*', 'AUU': '*'}\n"
     ]
    }
   ],
   "source": [
    "# import necessary modules\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from codonUtils import utils\n",
    "from codonTable import codonTable\n",
    "from queue import Queue as queue\n",
    "from random import choice\n",
    "class ffgenBC():\n",
    "    ''' A class used to generate fast fail codon tables with less than 20 Amino Acids\n",
    "    '''\n",
    "\n",
    "    def __init__(self):\n",
    "        return\n",
    "\n",
    "    @staticmethod\n",
    "    def triplet(knockouts = 0):\n",
    "        '''\n",
    "        A function used to generate triplet decoding, fast fail genetic codes\n",
    "        using a rational, 'top down' approach. Snakes along codon table to fill\n",
    "        16 maximally distant codons first (all greater than one mutation from\n",
    "        each other), then randomly chooses 4 codons to place the remaining\n",
    "        amino acids'''\n",
    "\n",
    "        ############################\n",
    "        # fill out first 16 codons #\n",
    "        ############################\n",
    "\n",
    "        # declare caching variables\n",
    "        usedCodons = set()\n",
    "        unusedAA = set(utils.residues[:-1])\n",
    "        # declare budding table\n",
    "        table = {}\n",
    "        # randomly permute rNTPs for positions 1 and 2, store 3 unpermuted\n",
    "        pos1 = np.random.permutation(utils.rNTPs)\n",
    "        pos2 = np.random.permutation(utils.rNTPs)\n",
    "        pos3 = utils.rNTPs\n",
    "        # store nucleotides for positions 1 and 2 in queues\n",
    "        queue1 = queue()\n",
    "        queue2 = queue()\n",
    "        for nt1, nt2 in zip(pos1, pos2):\n",
    "            queue1.put(nt1)\n",
    "            queue2.put(nt2)\n",
    "\n",
    "        # place the first 16 elements\n",
    "        for i in range(queue1.qsize()):\n",
    "            # get first nucleotide of next codon\n",
    "            nt1 = queue1.get()\n",
    "            for j in range(queue2.qsize()):\n",
    "                # get second nucleotide of next codon\n",
    "                nt2 = queue2.get()\n",
    "                # get third nucleotide\n",
    "                nt3 = pos3[j]\n",
    "                # build codon, assign it a residue\n",
    "                codon = nt1 + nt2 + nt3\n",
    "                AA = choice(tuple(unusedAA))\n",
    "                table[codon] = AA\n",
    "                # update caching variables\n",
    "                usedCodons.add(codon)\n",
    "                unusedAA.remove(AA)\n",
    "                # re-enqueue second nucleotide\n",
    "                queue2.put(nt2)\n",
    "            # re-enqueue first nucleotide\n",
    "            queue1.put(nt1)\n",
    "\n",
    "        # place the last 4 elements at least 2 substitutions away\n",
    "        availableCodons = set(utils.tripletCodons) - usedCodons \n",
    "        for i in range(len(unusedAA)-knockouts):\n",
    "            # pick a codon from the available set and assign it an amino acid\n",
    "            codon = choice(tuple(availableCodons))\n",
    "            AA = choice(tuple(unusedAA))\n",
    "            table[codon] = AA\n",
    "            # update cache variables\n",
    "            availableCodons = ffgenBC.updateAvailable3(codon, availableCodons)\n",
    "            usedCodons.add(codon)\n",
    "            unusedAA.remove(AA)\n",
    "\n",
    "        # assign unused codons to STOP\n",
    "        remainingCodons = set(utils.tripletCodons) - usedCodons\n",
    "        for codon in remainingCodons:\n",
    "            table[codon] = '*'\n",
    "\n",
    "        # return built table\n",
    "        return table\n",
    "\n",
    "    @staticmethod\n",
    "    def updateAvailable3(newCodon, availableSet):\n",
    "        ''' A static method used to update the set of codons that can be used\n",
    "        for triplet decoding fast fail code, given that a new codon is\n",
    "        occupied.''' \n",
    "\n",
    "        # iterate over remaining codons\n",
    "        copySet = list(availableSet)\n",
    "        for codon in copySet:\n",
    "            # remove codons that have two nucleotide overlaps\n",
    "            count = 0\n",
    "            for i in range(len(newCodon)):\n",
    "                count += (codon[i] == newCodon[i])\n",
    "            if count >= 2:\n",
    "                availableSet.remove(codon)\n",
    "        # return updated set\n",
    "        return availableSet\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    # availableCodons = set(utils.tripletCodons)\n",
    "    # count = 0\n",
    "    # while len(availableCodons) > 0:\n",
    "    #     codon = random.choice(tuple(availableCodons))\n",
    "    #     availableCodons = ffgenBC.updateAvailable3(codon, availableCodons)\n",
    "    #     count += 1\n",
    "    print(ffgenBC.triplet())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
